### PR TITLE
chore: create basic layout components and associated utils

### LINKED ISSUE #27000: [Task]: Create LayoutComponents HOC.
1. Create an HOC called Layout Component (LC).
2. LC wraps all children in FlexLayout.
3. if `props.isDropTarget`, LC additionally renders the Dragging Arena for Anvil, inside Flex Layout.
4. if `props.rendersWidgets` LC uses WidgetFactory.create to render child widgets; else it uses LayoutFactory to render child layouts.
5. Create a util method to expose renderChild (WidgetFactory.create) functionality.

### LINKED ISSUE #27001: [Task]: Create basic layouts.
1. Create the following Layouts:
a) AlignedColumn. - [layoutType: 'ALIGNED_COLUMN]
b) AlignedRow. - [ALIGNED_ROW]
c) Column. - [COLUMN]
d) Row. - [ROW]
2. Each layout should use LayoutComponent HOC to render children layouts / widgets.
3. Each layout should expose the following static methods:
a) `getWidth => (args): number`
b) `getHeight => (args): number`
c) `getChildTemplate => (args): LayoutComponentProps | undefined`
d) `addChild => (args): LayoutComponentProps`
e) `removeChild => (args): LayoutComponentProps`
f) `deriveHighlights => (args): HighlightInfo[]`

### LINKED ISSUE #27002: [Task]: Create LayoutFactory and LayoutUtils.
1.`LayoutFactory`:
a) Returns appropriate LayoutComponent based on `layoutType`.
2. `LayoutUtils`:
a) `renderLayouts()`: Iterates through a layout tree, rendering all nested layouts / widgets in a depth first manner. Uses LayoutFactory and WidgetFactory to accomplish this.

### PR DESCRIPTION
**Pull request by prsidhu**
## Description

Objective - Create scaffolding for layout components (LC).

- Created ``LayoutComponentHOC`` that will wrap the ``FlexLayout`` component and provide all the necessary functionality that is expected of an LC.
- Created ``LayoutFactory`` to initial available LCs.
- Created base LCs that will be used to construct presets - ``AlignedColumn, AlignedRow, Column and Row``.
- Created various utils to manage functionality of LCs.
- Added unit tests for utils.

#### PR fixes following issue(s)
Fixes #27000
Fixes #27001 
Fixes #27002 

- New feature (non-breaking change which adds functionality)

## Testing

- [x] Manual
- [ ] JUnit
- [x] Jest
- [ ] Cypress

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Comment by marks0351** [comment]
Some points of discussion wrt to code design:
To enable layout components inside the existing canvas flow i.e, the renderAppsmithCanvas we would need a Parent Component per each renderMode that would kick start rendering of layout components with the `layout` prop. 

`AnvilLayoutEditorCanvas` and `AnvilLayoutViewerCanvas`.
in both components an entry point is needed.

I see that renderLayouts is the kick stater which iterates through `layout` data of a canvas and gets the layout component based on type and renders it.
each layout component is wrapped with a hoc that adds a DnD layer and takes the decision whether the component should render layouts or render its own implementation of renderChildWidgets

Registration of layout components like widgets has hurdles coz we wont have access to app state during the registration cycle so doing it on-demand would not be possible with todays arch. I would recommend we skip the same widget registration flow and instead just have a constant mapping layout type and the component.

for the entry point mentioned above for AnvilLayoutEditorCanvas, Would it make sense to use a parent Component `LayoutComponetsMaker`, to pick the layout component from a constant which would be a map of all layout components which is mentioned in the `layout` property and render it instead of picking it from a layout registry that adds a HOC for every widget.

regarding the static functions per layout component, if they are to DRY the components, 
I would suggest we use an Abstract base class for Layout components(`BaseLayoutComponent`) just like BaseWidget and extend them per component with their own definitions.

The `LayoutComponentHOC` currently adds a DnD layer and the decision to renderLayouts or widgets, we can move this as a common logic inside the abstract class(`BaseLayoutComponent`) for layout components.
A bit skeptical about the effect of Object.assign in a HOC.

one more concern was that I feel a `LayoutComponent` is trying to do a lot by being a renderer of layout skeleton as well as being a renderer of children in a particular layout manner. there by the prop `layout` has multiple type definitions, I think we should isolate them, like have layout skeleton renderers and layout with widget renderers. and have appropriate type definitions and components for each. and whenever we see an opportunity to dry them we have tools like hooks, hocs and utils to do that.

@prsidhu @riodeuno let me know ur thoughts.

File: app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * that are rendered by this layout or its child layouts.
+ * @param layoutProps | LayoutComponentProps
+ * @param map | Record<string, WidgetProps>
+ * @param res | Record<string, WidgetProps>
+ * @returns Record<string, WidgetProps>
+ */
+export function getChildrenMap(
+  layoutProps: LayoutComponentProps,
+  map: LayoutComponentProps["childrenMap"],
+  res: LayoutComponentProps["childrenMap"] = {},
```
Comment (riodeuno) [inline app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:57]: It seems like this third argument is the return value. Is there a reason why this exists as an argument?

Comment (riodeuno) [reply app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:57]: Please ignore this. Just saw the recursion in the function

File: app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  if (!layout || !layout.length) return res;
+
+  // Parse each item of layout.
+  for (const each of layout) {
+    // if type === string => each is a widgetId.
+    if (typeof each === "string") {
+      // add widget to the resultant map.
+      res[each] = map[each];
+    } else if (Array.isArray(each)) {
+      // if type === array => each could be a list of widgetIDs.
```
Comment (riodeuno) [inline app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:70]: Is this assumption valid? What if we have deeply nested layouts?

File: app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+export function renderWidgetsInAlignedRow(
+  props: LayoutComponentProps,
+  childrenWidgetIds: string[],
+): React.ReactNode {
+  const { canvasId, childrenMap, layoutId, renderMode } = props;
+  // check if layout renders a Fill widget.
+  const hasFillWidget: boolean = isFillWidgetPresentInList(
+    Object.values(childrenMap || {}),
+  );
+  const mode: RenderMode = renderMode || RenderModes.CANVAS;
```
Comment (riodeuno) [inline app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:113]: Once again, I believe `renderMode` shouldn't be a part of props here in the system. Ideally, this should be a modal property of a higher context of the application.

**Review by riodeuno — COMMENTED** [review]
I have reviewed much of the code, but not all of it. Will complete it by tomorrow.

File: app/client/src/layoutSystems/anvil/utils/anvilTypes.ts [hunk 4]
```diff
@@ -1,12 +1,56 @@
...
+export type LayoutComponentType =
+  | "ALIGNED_COLUMN"
+  | "ALIGNED_ROW"
+  | "COLUMN"
+  | "ROW";
+
 export interface LayoutComponentProps {
   layout: LayoutComponentProps[] | string[] | string[][]; // Array of layout components or widgets to render.
   layoutId: string; // Identifier of layout
   layoutStyle?: { [key: string]: any }; // React.CSSProperties for overriding default layout style.
```
[the remark this answers is not in the document]
Comment (prsidhu) [reply app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:14]: @riodeuno yes, CSSProperties doesn't cover all scenarios. We also have to account for responsive values that are entertained by the container queries implementation on WDS.

File: app/client/src/layoutSystems/anvil/utils/anvilTypes.ts [hunk 5]
```diff
@@ -1,12 +1,56 @@
...
-  layoutType: string; // Used to identify the correct layout component to render.
+  layoutType: LayoutComponentType; // Used to identify the correct layout component to render.
 
   allowedWidgetTypes?: string[]; // Array of widget types that can be dropped on the layout component.
```
[the remark this answers is not in the document]
Comment (prsidhu) [reply app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:17]: empty array also signifies absence of a rule, which does the same job.

File: app/client/src/layoutSystems/anvil/utils/anvilTypes.ts [hunk 6]
```diff
@@ -1,12 +1,56 @@
...
+export interface LayoutComponent extends React.FC<LayoutComponentProps> {
+  // add all other static props here
+  type: LayoutComponentType;
+  // Add a child widget / layout to the parent layout component.
+  addChild: (
+    props: LayoutComponentProps,
+    children: string[] | LayoutComponentProps[],
+    highlight: HighlightInfo,
+  ) => LayoutComponentProps;
+  getWidth: (arg0: any) => number;
```
[the remark this answers is not in the document]
Comment (prsidhu) [reply app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:38]: @riodeuno @marks0351 Final definition and requirement of this function will become clear after the highlights task is done. I'll remove it if it doesn't serve a purpose anymore.

File: app/client/src/layoutSystems/anvil/utils/anvilTypes.ts [hunk 7]
```diff
@@ -1,12 +1,56 @@
...
+
+  childrenMap?: Record<string, WidgetProps>; // Map of child widget ids to their props.
+  renderMode?: RenderMode;
+}
+
+export interface LayoutComponent extends React.FC<LayoutComponentProps> {
+  // add all other static props here
+  type: LayoutComponentType;
+  // Add a child widget / layout to the parent layout component.
+  addChild: (
```
[the remark this answers is not in the document]
Comment (prsidhu) [reply app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:40]: @riodeuno child can be either widget or layout. ``HighlightInfo`` provides the drop information that is used by the layout to place the child.

File: app/client/src/layoutSystems/anvil/utils/layouts/layoutUtils.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+export function addChildToLayout(
+  props: LayoutComponentProps,
+  children: string[] | LayoutComponentProps[],
+  highlight: HighlightInfo,
+): LayoutComponentProps {
+  const layout: any = props.layout;
+  const { rowIndex: index } = highlight;
+  return {
+    ...props,
+    layout: [...layout.slice(0, index), ...children, ...layout.slice(index)],
```
[the remark this answers is not in the document]
Comment (prsidhu) [reply app/client/src/layoutSystems/anvil/utils/layouts/layoutUtils.ts:36]: @riodeuno additionally, the layout will have to sort the children based on the said ``index`` for every render, as well as highlights calculation.

**Comment by prsidhu** [comment]
@riodeuno @marks0351 I have added the canvas implementation to this PR, along with minor review suggestions. 

I will include the larger suggestions in the anvil sagas PR as it will allow me to make the desired changes across the board in a single attempt.

Additionally, I've added three tasks to reconsider epic based on the review suggestions.

**Comment by prsidhu** [comment]
Closing this. The changes have been merged to release via #27980

**Pull request closed without merging**